Fix Script Editor saves C# files as embedded scripts.

nongvantinh opened the pull request:
fixes: https://github.com/godotengine/godot/issues/88543
fixes: https://github.com/godotengine/godot/issues/88160

Because of the limitations of compiled programming languages like C#, when a newly created *.cs file hasn't been compiled, we don't have any information about its `Path` or `Type` in the `assembly`. This means we end up creating an invalid instance of this file whenever there's a request. Consequently, multiple instances of the script can exist. When a new instance takes over the path, it clears the `path_cache` of the previous instance, leading to undefined behavior.

KoBeWi commented:
If other instances of the script exist, taking over path will make them unlinked from the original resource. It might cause attached scripts to stop updating (relevant for tool scripts).

KoBeWi commented:
So I just realized

> During the ResourceLoader::load process, the C# side creates a different instance and clears the path_cache of the old one, causing it to lose its path.

This points to a bug in C# loader, because CACHE_MODE_IGNORE is not supposed to do that.

KoBeWi commented:
Tested this PR and I'm getting
```
C:\godot_source\modules/gdscript/gdscript.cpp:702 - Condition "!p_keep_state && has_instances" is true. Returning: ERR_ALREADY_IN_USE
```
when modifying a GDScript file externally. Specifically it's a tool script attached in the current scene. After the error prints once, the script no longer updates.

nongvantinh committed:
Fix Script Editor saves C# files as embedded scripts

fixes: #88543
fixes: #88160

Because of the limitations of compiled programming languages like C#, when a newly created *.cs file hasn't been compiled, we don't have any information about its `Path` or `Type` in the `assemply`. This means we end up creating an invalid instance of this file whenever there's a request. Consequently, multiple instances of the script can exist. When a new instance takes over the path, it clears the `path_cache` of the previous instance, leading to undefined behavior.

nongvantinh commented:
Thanks for pointing this out, the new PR should correctly and safely fix the bug and any future bugs related to multiple instances of C# script.

KoBeWi commented on modules/mono/csharp_script.cpp:
```diff
@@ -2851,7 +2851,11 @@ Ref<Resource> ResourceFormatLoaderCSharpScript::load(const String &p_path, const
 
-	scr->set_path(p_original_path, true);
+	Ref<Resource> existing = ResourceCache::get_ref(p_path);
+	if (!existing.is_valid()) {
```
```suggestion
	if (existing.is_null()) {
```

KoBeWi commented:
While this works, it's not exactly how it should be implemented.
The code you added should take effect only when `p_cache_mode` is `CACHE_MODE_IGNORE`. Otherwise the new script should normally take over path.

nongvantinh commented:
Thanks. I was completely ignored the `p_cache_mode` parameter since it wasn't being used before. Now, I have modified it according to [the latest document specifications](https://docs.godotengine.org/en/latest/classes/class_resourceloader.html#:~:text=via%20load_threaded_get.-,enum%20CacheMode%3A,-CacheMode%20CACHE_MODE_IGNORE%20%3D). Please help me test it thoroughly.

akien-mga commented:
Thanks!
